examples/udp-echo.rs

```diff
@@ -0,0 +1,259 @@
+    }
+
+    let gateways = adapter.get_gateways()?;
+    println!("adapter gateways: {gateways:?}");
+
+    // adapter.set_name("MyNewName")?;
+    // println!("adapter name: {}", adapter.get_name()?);
+
+    // adapter.set_address("10.28.13.2".parse()?)?;
+
+    let session = Arc::new(adapter.start_session(wintun::MAX_RING_CAPACITY)?);
+    let reader_session = session.clone();
+    let writer_session = session.clone();
+
+    let (tx, rx) = channel::<NaiveUdpPacket>();
+
+    // Global flag to stop the session
+    static RUNNING: AtomicBool = AtomicBool::new(true);
+
+    let reader = std::thread::spawn(move || {
+        let block = || {
+            while RUNNING.load(Ordering::Relaxed) {
+                let packet = reader_session.receive_blocking()?;
+                // recieved IP packet
+                let bytes = packet.bytes();
+
+                let udp_packet = extract_udp_packet(bytes);
+                if let Err(err) = udp_packet {
+                    println!("{}", err);
+                    continue;
+                }
+
+                // swap src and dst
+                let mut udp_packet = udp_packet?;
+                let src_addr = udp_packet.src_addr;
+                let dst_addr = udp_packet.dst_addr;
+                udp_packet.src_addr = dst_addr;
+                udp_packet.dst_addr = src_addr;
+
+                // send to writer
+                tx.send(udp_packet)?;
+            }
+            Ok::<(), Box<dyn std::error::Error>>(())
+        };
+        if let Err(err) = block() {
+            println!("Reader {}", err);
+        }
+    });
+
+    let writer = std::thread::spawn(move || {
+        let block = || {
+            while RUNNING.load(Ordering::Relaxed) {
+                let resp = rx.recv()?;
+
+                let src_addr = match resp.src_addr.ip() {
+                    IpAddr::V4(addr) => addr,
+                    IpAddr::V6(_) => return Err("IPv6 addresses are not supported".into()),
+                };
+
+                let dst_addr = match resp.dst_addr.ip() {
+                    IpAddr::V4(addr) => addr,
+                    IpAddr::V6(_) => return Err("IPv6 addresses are not supported".into()),
+                };
+
+                let v = generate_random_bytes(2)?;
+                let id = u16::from_ne_bytes([v[0], v[1]]);
+
+                // build response IP packet
+                use packet::Builder;
+                let ip_packet = packet::ip::v4::Builder::default()
+                    .id(id)?
+                    .ttl(64)?
+                    .source(src_addr)?
+                    .destination(dst_addr)?
+                    .udp()?
+                    .source(resp.src_addr.port())?
+                    .destination(resp.dst_addr.port())?
+                    .payload(&resp.data)?
+                    .build()?;
+
+                // // The following code will be better than above, the `ipv4_udp_build` function link is
```

Comment:
Why can't we use this one?

Reply:
Because we have to add a file `protocol.rs`, it's unnecessary.  When I use `packet` crate,  I send an even-length UDP packet, the data is correct. But when I send an odd-length UDP packet, the data packet is wrong. Therefore the correct approach is to submit a patch to the `packet` crate.